### ISSUE TITLE
Better handle opening and closing subscriptions

### ISSUE THREAD

**Issue by Urigo** [Urigo/angular-meteor#304]
We want to help developers handle subscriptions more easily.

We have internal discussions about that and also many people from the community started coming up with interesting solutions.

I'm opening this issue so that we will move the discussion here so everyone can affect the solution.

Starting point:

Right now there are 2 general approaches with different solutions for each of them:
1. Router level
2. Controller level

I personally tend to go with the controller level as it's a more general solution (closer to component level subscription like Blaze and Angular 2.0) but it has it's problems and I'm not sure what the actual technical solution even be.
In routers I've thought and also heard many solutions so it might be easier to implement.

Looking forward for your thoughts

**Comment by revov** [comment]
@Urigo, what works for me is a simple Factory that I use in my project to help me with the huge overhead of cleaning up subscriptions, collections and objects:

```javascript
angular.module('myModule')
    .factory('meteorWrapperFactory', [
        '$meteor', '$q',
        function($meteor, $q) {
            return {
                subscribe: function( scope, subscriptionParams /*variadic*/ ) {
                    var deferred = $q.defer();
                    var args = _.takeRight(arguments, arguments.length-1);
                    args.push({
                        onReady: function () {
                            deferred.resolve(subscriptionHandle);
                        },
                        onError: deferred.reject
                    });

                    var subscriptionHandle = Meteor.subscribe.apply( this, args );
                    scope.$on('$destroy', function() {
                        subscriptionHandle.stop();
                    });

                    return deferred.promise;
                },
                collection: function( scope, collectionParams /*variadic*/ ) {
                    var collection = $meteor.collection.apply( this, _.takeRight(arguments, arguments.length-1) );
                    scope.$on('$destroy', function() {
                        collection.stop();
                    });

                    return collection;
                },
                object: function( scope, objectParams /*variadic*/ ) {
                    var object = $meteor.object.apply( this, _.takeRight(arguments, arguments.length-1) );
                    scope.$on('$destroy', function() {
                        object.stop();
                    });

                    return object;
                }
            };
        }
    ]);
```

Note that I am using lodash to get the variadic array of arguments, but this can easily be achieved with vanilla Javascript (considering Meteor's deprecation of Underscore js). This is a really simple solution that won't fit all needs, but it gets rid of the awful cleanup overhead while still preserving performance and providing flexibility and ease of use. What I like about this approach is that it doesn't enforce architectural decisions on the developer. It just encapsulates the tedious boilerplate code.

Personally, I use the factory in my controllers and directives like this:
```javascript
var mySubscription = meteorWrapperFactory.subscribe( $scope, 'mySubscription' );
mySubscription
    .then(function() {
        $scope.collection1 = meteorWrapperFactory.collection( $scope, Collection1, false );
        $scope.collection2 = meteorWrapperFactory.collection( $scope, Collection2, false );
    });
```

I find Meteor's concept about subscriptions and collections reasonable enough and I believe that any further abstraction would break the separation of concerns. Anyway, I am interested to see what solutions will be proposed.

**Comment by auterium** [comment]
@revov I really like your approach as it's a self-contained, non-invasive one.

I did a different approach myself, though I did it within angular-meteor project. I've created the pull request with my proposal. I added the example of using it with $state, as is the way I usually work, but it can also be used with routes or any other method.

What do you guys think of a combination?
```javascript
angularMeteorSubscribe.service('$meteorSubscribe', ['$q',
  function ($q) {
    var self this;

    this.subscribe = function() {/* ommited for comment */};

    this.subscribeSelfDestruct = function() {
      var args = Array.prototype.slice.call(arguments);
      var $scope = args.shift();

      if(!$scope.$on) {
        console.warn('$meteor.subscribeSelfDestruct expects $scope as the first param!');
        return;
      }

      return self.subscribe.apply(self.subscribe, args).then(function(handle) {
        $scope.$on('$destroy', function() {
          handle.stop();
        });
      });
    };
}]);
```

Of course it would be applied for collection & object too

**Comment by netanelgilad** [comment]
There are a few things we need to handle when handling subscriptions.
 1. Getting the handle only from the ready promise is not good. That means that you get the handle only after the initial load of the subscription happened. What if the developer wants to stop the subscription before the initial load? That cannot be achieved right now. To solve this, I think we need to return the handle as a property on the returned promise, or return an object with 2 properties: ready (the promise) and handle (the subscription handle).
 2. All of the solutions here stopped the subscriptions on the `$destroy` event of the $scope. I think that instead of wrappers we can actually add methods to the $scope object like we did with getReactively and that way the use of the functions would be easier: `$scope.subscribe('sub')`.
 3. I still think a solution for subscriptions in the resolve part of $state is needed. I am not quite sure what is the way to do this yet, but I would love to see ideas in that area.

**Comment by smoke** [comment]
I really like @revov solution - it feels somehow complete and @netanelgilad comments are all valid - the subscribtionHandle could be of use just in case.

Also I just want to stress on a point to consider - subscriptions could be used in services, where there is no scope, so the option to subscribe should also support such cases.

**Comment by Urigo** [comment]
@revov really nice solution!
I like when it is binded to the controller.
Question - why are you closing the collections?  I haven't seen something about that in Meteor's docs?

I've took your solution and @netanelgilad notes about adding it directly to the scope.
Here is my initial suggestion: https://github.com/Urigo/angular-meteor/commit/4315a6add8ffbce2dd730764288aa7c1f4dbe37d

It's still annoying that I almost copied the entire $meteor.subscribe service and haven't reused the code, would love your thoughts on that.

@auterium and @smoke 
As placing it in controller is closer to Meteor's template.subscribe and more general I want to solve that first, but I agree we should provide also something to the routers and services.

Let's start with closing controllers - what do you think about my suggestion?


@netanelgilad check please review my use of 'this' ....

**Comment by smoke** [comment]
> Question - why are you closing the collections? I haven't seen something about that in Meteor's docs?

This is needed to free up resources (memory, CPU, bandwidth) of objects and maintaining them when they are not needed any more. Such usage is rarely the point of tutorials, so they omit it for the sake of keeping it simple, but this is really important thing in production applications.

Let me try and explain it with an example - lets assume you have page that lists details of a Party and Users going to it. This would be a subscription for a publisher that pulls this specific data, once you navigate out of that page to another Party, the old information is completely unneeded and would not free the resources unless stopped. In this case you would get memory leak, that the more user use your application, the more resources it would eat and once the limits are reached a crash is expected.

Hope above explains is nicely.

**Comment by netanelgilad** [comment]
Just to be a bit more specific on @smoke answer, if the collection is not stopped when navigating away the app will leak observe handles (on every $meteor.collection created). It will not leak documents from the collection (that is handled when you stop a subscription).

@smoke @revov  that is a good catch that we would have forgotten about :beers:

**Comment by revov** [comment]
Yes, stopping subscriptions is equally important to stopping collections and objects.

However, I don't like the solution with attaching the subscribe function to the $rootScope. What if we are in a **directive's isolate scope**? We won't be able to use this utility in such cases even though we have a valid scope object because the isolate scope does not inherit the properties of the $rootScope. This use case is really common and a well abstracted angular meteor application will make use of subscriptions/collections/objects inside directives quite often.

On the other topic - getting the subscription handle before the promise is resolved - I don't see the use cases of this. The subscription handle is useless before the subscription is marked as ready except for the cases where we want to stop the subscription. With the approach we are discussing here this is done automatically on scope destruction so cases where we need to manually stop a subscription should be rare and thus require using the Meteor API directly. Moreover, if we want to stop a subscription before it is ready just so that we can resubscribe with new parameters - we have the **autorun** method doing exactly this - stopping all subscriptions invoked inside its computation when it is invalidated.

On the topic of using subscriptions in services - there is nothing we can do to automate this. Developers should implement their own logic and use the lower level APIs.

**Comment by netanelgilad** [comment]
@Urigo We need to use the prototype solution so the function would be available on isolate scopes as mentioned by @revov (same as getReactively).
@revov on the matter of closing the subscription - what you say is indeed correct, but the problem is that in those rare situations where you do need to close the subscription before it is loaded, you can't do it even with Meteor API, only through the subscription handle (as far as I know...)

**Comment by Urigo** [comment]
@revov @netanelgilad I've update the branch so that it will use Object.prototyepOf so we will support also isolated scope:
https://github.com/Urigo/angular-meteor/commit/1be43280b35bb891e609ff989e99fa44473524ae
What do you think?

**Comment by auterium** [comment]
@Urigo I like the idea of using Object.prototypeOf, but shouldn't this be a wrapper of $meteor.subscribe? The code for handling the subscription is already there, and this is just a higher level api method.

**Comment by Urigo** [comment]
@auterium I agree, but I have to get the subscription handle not from the promise.
Would love your suggestions about how to do it better to minimize the code duplication

**Comment by auterium** [comment]
@Urigo what about a low level subscribe method that returns the subscription directly instead of a promise?

Warning: Untested code

```javascript
angularMeteorSubscribe.service('$meteorSubscribe', ['$q', function($q) {
  var self = this;

  this._subscribe = function(scope, deferred, args) {
    var subscription = null;

    args.push({
      onReady: function() {
        deferred.resolve(subscription);
      },
      onError: function(err) {
        deferred.reject(err);
      }
    });

    subscription =  Meteor.subscribe.apply(scope, args);

    return subscription;
  };

  this.subscribe = function() {
    var deferred = $q.defer();
    var args = Array.prototype.slice.call(arguments);

    self._subscribe(this, deferred, args);

    return deferred.promise;
  };
}]);
```

```javascript
angularMeteorSubscribe.run(['$rootScope', '$q', '$meteorSubscribe', function($rootScope, $q, $meteorSubscribe) {
  Object.getPrototypeOf($rootScope).subscribe = function() {
    var deferred = $q.defer();
    var args = Array.prototype.slice.call(arguments);

    var subscription = $meteorSubscribe._subscribe(this, deferred, args);

    this.$on('$destroy', function() {
      subscription.stop();
    });

    return deferred.promise;
  };
}]);
```

**Comment by Urigo** [comment]
@auterium oops just saw that.
that sounds great, maybe you want to pull request???

**Comment by Urigo** [comment]
now that we took care of scope subscription we still need to solve (or not) 2 other things that came up here:

1. stopping observes of $meteor.collection and $meteor.object - how would the solution look?  like $scope.subscribe?
2. Do something for routes - now that we have it on the controllers, do we still need this?

**Comment by auterium** [comment]
@Urigo thinking of the $meteor.object & $meteor.collection to work similar to $scope.subscribe could create some collisions if someone would like to set params to $scope that have such names. I think the better way to declare such methods should be prefixing $meteor, as it's not part of the $scope's api, but an addon. It would then look like this:
- $scope.$meteorSubscribe
- $scope.$meteorCollection
- $scope.$meteorObject
All of the above would work as stated in the code above.

I would definitelly like to have something for the routes as when I started using meteor was with iron router and loved that you only had to declare that it had to wait for the subscriptions and it would then handle the closing when changing routes.

If we do both solutions, then we should separate them into packages as both solve the same problem but with different approaches and there's no need to have both on a given project.

**Comment by Urigo** [comment]
@auterium that's a good idea.
As I'm working on the ng-vegas talk this week, I would love so pull requests (and packages) on all of those.
The start is already there in $scope.susbcribe

**Comment by auterium** [comment]
@Urigo I've been researching about having the subscriptions on the routes using the resolve property and here are my findings:
- It works as desired, but it slows down the route change as it waits until the subscription is resolved; this is even worse if you have multiple subscriptions declared on the resolve part.
- In order to do a single item resolution (i.e. display/edit page of a single doc) you can do resolve chainging:
```javascript
.state('party_edit', {
  url: '/parties/:id',
  resolve: {
    singlePartySubscription: ['$meteor', '$stateParams', function($meteor, $stateParams) {
      return $meteor.subscribe('single-party', $stateParams.id;
    },
    party: ['$meteor', '$stateParams', '$q', 'singlePartySubscription', function($meteor, $stateParams, $q) {
      // This will not be executed until the singlePartySubscription is resolved
      var party = $meteor.object(Parties, $stateParams.id);

      if(!party._id) {
        return $q.defer().reject('NOT_FOUND');
      }

      return party;
    }]
  }
});
```
- By adding a listener to $stateChangeSuccess, we can track & stop resolved values that can be stopped:
```javascript
.run(['$rootScope', '$state', function($rootScope, $state) {
  $rootScope.$on('$stateChangeSuccess', function(event, toState, toParams, fromState, fromParams) {
    toState.$stoppables = [];

    angular.forEach($state.$current.locals.resolve.$$values, function(resolvedValue) {
      if(resolvedValue.stop) {
        toState.$stoppables.push(resolvedValue);
      }
    });

    if(!fromState.$stoppables) {
      return;
    }

    var stoppable;
    while(stoppable = fromState.$stoppables.shift()) {
      stoppable.stop();
    }
  });
}]);
```
- Subscriptions that are created within a $meteor.autorun() don't need to be stopped as the autorun tracks them and stops them when invalidating the previous autorun.

**Comment by auterium** [comment]
One thing worth documenting is that I used Meteor.default_connection._subscriptions to debug and check for active subscriptions while testing subscription handling.

**Comment by Urigo** [comment]
ok, closing as we have all that in.

About routes, looks like the code example might be enough or add something as a separate package.

Please add things to the docs if you think it is worthwhile


thank you so much everyone for those great features and collaborations!
